lib/Dialect/Openfhe/Transforms/ConfigureCryptoContext.cpp

```diff
@@ -0,0 +1,161 @@
+bool hasMulOp(func::FuncOp op) {
+  bool result = false;
+  op.walk<WalkOrder::PreOrder>([&](Operation *op) {
+    if (isa<openfhe::MulOp>(op) || isa<openfhe::MulNoRelinOp>(op) ||
+        isa<openfhe::MulPlainOp>(op)) {
+      result = true;
+      return WalkResult::interrupt();
+    }
+    return WalkResult::advance();
+  });
+  return result;
+}
+
+// Helper function to find all the rotation indices in the function
+SmallVector<int64_t> findAllRotIndices(func::FuncOp op) {
+  SmallVector<int64_t> rotIndices;
+  op.walk([&](openfhe::RotOp rotOp) {
+    // TODO : I'm not sure if this is the right way to get the index
+    auto typedIndex =
+        dyn_cast<arith::ConstantOp>(rotOp.getIndex().getDefiningOp())
+            .getValue();
+    int64_t rotIndex = dyn_cast<IntegerAttr>(typedIndex).getInt();
+    if (std::find(rotIndices.begin(), rotIndices.end(), rotIndex) ==
+        rotIndices.end()) {
+      rotIndices.push_back(rotIndex);
+    }
+    return WalkResult::advance();
+  });
+  return rotIndices;
+}
+
+// function that generates the crypto context with proper parameters
+LogicalResult generateGenFunc(func::FuncOp op, const std::string &genFuncName,
+                              ImplicitLocOpBuilder &builder) {
+  Type openfheContextType =
+      openfhe::CryptoContextType::get(builder.getContext());
+  SmallVector<Type> funcArgTypes;
+  SmallVector<Type> funcResultTypes;
+  funcResultTypes.push_back(openfheContextType);
+
+  FunctionType genFuncType =
+      FunctionType::get(builder.getContext(), funcArgTypes, funcResultTypes);
+  auto genFuncOp = builder.create<func::FuncOp>(genFuncName, genFuncType);
+  builder.setInsertionPointToEnd(genFuncOp.addEntryBlock());
+
+  // TODO : Calculate the appropriate values by analyzing the function
+  int64_t mulDepth = 2;
+  int64_t plainMod = 4295294977;
+  Type openfheParamsType = openfhe::CCParamsType::get(builder.getContext());
+  Value ccParams = builder.create<openfhe::GenParamsOp>(openfheParamsType,
+                                                        mulDepth, plainMod);
+  Value cryptoContext =
+      builder.create<openfhe::GenContextOp>(openfheContextType, ccParams);
+
+  builder.create<func::ReturnOp>(cryptoContext);
+  return success();
+}
+
+// function that configures the crypto context with proper keygeneration
+LogicalResult generateConfigFunc(func::FuncOp op,
+                                 const std::string &configFuncName,
+                                 bool hasMulOp, SmallVector<int64_t> rotIndices,
+                                 ImplicitLocOpBuilder &builder) {
+  Type openfheContextType =
+      openfhe::CryptoContextType::get(builder.getContext());
+  Type privateKeyType = openfhe::PrivateKeyType::get(builder.getContext());
+
+  SmallVector<Type> funcArgTypes;
+  funcArgTypes.push_back(openfheContextType);
+  funcArgTypes.push_back(privateKeyType);
+
+  SmallVector<Type> funcResultTypes;
+  funcResultTypes.push_back(openfheContextType);
```

Comment:
just curious, but does this need to be returned from the configuration funct if it modifies the context arg?

Reply:
I think I might not have understood the question correctly. Could you elaborate, please?

Reply:
oh, it might be me not understanding! I *think* this function takes a crypto context and private key, and then generates the needed keys in the context. so if it modifies the context (i'm not sure if it modifies or copies), then i don't think it needs to return the context. so the user would do something like

```
OpenFHEContext context = __generate_crypto_context();
// create pub/priv key
__configure_crypto_context(context, priv_key);
```

Reply:
Oh, I see! You're right, it just modifies the context, not makes a new copy, so there's no need to return the context!

Though, again, currently it doesn't work since OpenFhePkeEmitter allows single value return only 😄 (Zero value return is not allowed)

Reply:
Ah got it! Adding it to that thread, and then I can help fix it up from there.